text/0000-restrict-constants-in-patterns.md

```diff
@@ -0,0 +1,568 @@
+it would be better to design them as a first-class feature, not adapt
+constants for that purpose.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+**Should we also adjust the exhaustiveness and match analysis
+algorithm to be more conservative around constants?** This RFC just
+proposes limiting the types of constants that can be used in a match
+pattern. However, since the code currently inlines the actual values
+of constants before doing exhaustiveness checking, this also implies
+that it can compute exhaustiveness and dead-code in cases where it
+arguably should not be able to.
+
+For example, the following code
+[fails to compile](http://is.gd/PJjNKl) because it contains dead-code:
+
+```rust
+const X: u64 = 0;
+const Y: u64 = 0;
+fn bar(foo: u64) {
+    match foo {
+        X => { }
+        Y => { }
+        _ => { }
+    }
+}
+```
+
+However, we would be unable to perform such an analysis in a more
+generic context, such as with an associated constant:
+
+```rust
+trait Trait {
+    const X: u64;
+    const Y: u64;
+}
+
+fn bar<T:Trait>(foo: u64) {
+    match foo {
+        T::X => { }
+        T::Y => { }
+        _ => { }
+    }
+}
+```
+
+Here, although it may well be that `T::X == T::Y`, we can't know for
+sure. So, for consistency, we may wish to treat all constants opaquely
+regardless of whether we are in a generic context or not.
+
+Another argument in favor of treating all constants opaquely is that
+the current behavior can leak details that perhaps were intended to be
+hidden. For example, imagine that I define a fn `hash` that, given a
+previous hash and a value, produces a new hash.  Because I am lazy and
+prototyping my system, I decide for now to just ignore the new value
+and pass the old hash through:
+
+```rust
+const fn add_to_hash(prev_hash: u64, _value: u64) -> u64 {
+    prev_hash
+}
+```
+
+Now I have some consumers of my library and they define a few constants:
+
+```rust
+const HASH_OF_ZERO: add_to_hash(0, 0);
+const HASH_OF_ONE: add_to_hash(0, 1);
+```
+
+And at some point they write a match statement:
+
+```rust
+fn process_hash(h: u64) {
+    match h {
+        HASH_OF_ZERO => /* do something */,
+        HASH_OF_ONE => /* do something else */,
+        _ => /* do something else again */,
+}
+```
+
+As before, what you get when you [compile this](http://is.gd/u5WtCo)
+is a dead-code error, because the compiler can see that `HASH_OF_ZERO`
+and `HASH_OF_ONE` are the same value.
+
+Part of the solution here might be making "unreachable patterns" a
+warning and not an error. The author feels this would be a good idea
+regardless (though not necessarily as part of this RFC). However,
+that's not a complete solution, since -- at least for `bool` constants
+-- the same issues arise if you consider exhaustiveness checking.
+
+On the other hand, it feels very silly for the compiler not to
+understand that `match some_bool { true => ..., false => ... }` is
```

Comment:
I'd say `true` and `false` feel more like unit variants of "`enum bool { false, true }`" than opaque constants.